src/sources/util/tcp/request_limiter.rs

```diff
@@ -0,0 +1,158 @@
+use std::cmp::Ordering;
+use std::sync::{Arc, Mutex};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+
+const EWMA_WEIGHT: f32 = 0.1;
+
+const MINIMUM_PERMITS: usize = 2;
+
+pub struct RequestLimiterPermit {
+    semaphore_permit: Option<OwnedSemaphorePermit>,
+    request_limiter_data: Arc<Mutex<RequestLimiterData>>,
+}
+
+impl RequestLimiterPermit {
+    pub fn decoding_finished(&self, num_events: usize) {
+        let mut request_limiter_data = self.request_limiter_data.lock().unwrap();
+        request_limiter_data.update_average(num_events);
+    }
+}
+
+impl Drop for RequestLimiterPermit {
+    fn drop(&mut self) {
+        if let Ok(mut request_limiter_data) = self.request_limiter_data.lock() {
+            let target = request_limiter_data.target_requests_in_flight();
+            let current = request_limiter_data.total_permits;
+
+            match target.cmp(&current) {
+                Ordering::Greater => request_limiter_data.increase_permits(),
+                Ordering::Equal => {
+                    // only release the current permit (when the inner permit is dropped automatically)
+                }
+                Ordering::Less => {
+                    let permit = self.semaphore_permit.take().unwrap();
+                    request_limiter_data.decrease_permits(permit);
+                }
+            }
```

Comment:
Have you considered introducing any hysteresis into this comparison? ie consider the values equal if they are within `N` of each other?

Reply:
I don't think that's necessary here. The range of values for the permits is pretty small, `[2, num_cpus]`, and I don't expect the average from the EWMA to vary much.